### ISSUE TITLE
Push notifications in PWA

### ISSUE THREAD

**Issue by startupgurukul** [ionic-team/capacitor#1282]
Folks thank you for capacitor.

Push notifications seem to be supported only in iOS and Android, but not Web, is there any reason for this?

Kindly enlighten.

**Comment by hsta** [comment]
If I would have to take a guess it is currently missing because of the required service worker. I guess a lot of projects already have a service worker. Our projects for example use the `ngsw-worker` from `@angular/pwa`, however, other projects may use totally different service workers so capacitor cannot assume if there is one and if so which one it is.

Like stated on https://capacitor.ionicframework.com/docs/basics/progressive-web-app/ writing service workers from scratch is complex. And "forcing" projects to use a certain one or even a new capacitor maintained service worker wouldn't be very good imho.

However, since I also would like to see Push Notification support for PWAs my suggestion would be to write a "smallest possible" service worker "extension", which only implements everything required for Push Notifications. Such a mini service worker could be included in the actual service worker file simply with `importScripts('./capacitor-push-notifications-sw.js');`. 
We have quite good experience with using just a 2 lines service worker file which first imports the native Angular service worker and then our custom service worker. This method is agnostic to which "main" service worker is used and updates of the main service worker or its build process. But it would still require some manual setup (adding the `importScripts` and maybe registering they new "container" service worker). Still I think it could be worth it.

**Comment by mlynch** [comment]
Agreed, we need to figure out a good solution for service workers (likely what Ionic does: a basic implementation using workbox) and then add push support. Marking this as high priority.

**Comment by CFT-Chris** [comment]
@mlynch Is there any branch currently that has development started on this task? I could probably help contribute to get this out sooner since I'm dealing with this now in an Ionic+Capacitor project.

**Comment by CFT-Chris** [comment]
@mlynch @jcesarmobile, any chance this issue and the PR I submitted be discussed soon?

**Comment by jmdavid** [comment]
Ionic team, please proceed as all tests passed and no conflicts.

**Comment by tvvignesh** [comment]
@mlynch Since this is still open, I am moving ahead using Workbox separate from Capacitor. In case this gets implemented, I hope the migration process would be smooth for me. Thanks.

**Comment by dthib** [comment]
Hi Capacitor team,
Thanks for the great work. Any news on this topic ?
Best,

**Comment by reslear** [comment]
Any news?

**Comment by spotvin42** [comment]
Hello, thanks everyone for the work on Capacitor. Just wondering whether this is coming soon?

**Comment by hardmaster92** [comment]
If you use Angular, for now you can use their service worker package

**Comment by sandstrom** [comment]
Most probably know already, but Web Push Notifications is coming to Safari (both iOS and macOS).

https://webkit.org/blog/12824/news-from-wwdc-webkit-features-in-safari-16-beta/#web-push-for-macos

**Comment by hardmaster92** [comment]
> Most probably know already, but Web Push Notifications is coming to Safari (both iOS and macOS).
> 
> https://webkit.org/blog/12824/news-from-wwdc-webkit-features-in-safari-16-beta/#web-push-for-macos
> 
> _(I don't know how that would work in a webview app wrapped within a native shell though, which is how most Capacitor apps are packaged on iOS)_

They have already been available for macOS (I've even implemented them in 2019), it's just a bit different. It's only iOS where they haven't been available.

**Comment by strarsis** [comment]
+1, I want to use it in Ionic as all the other components!

**Comment by strarsis** [comment]
iOS/iPad OS/macOS Safari support for browser push notifications finally rolled out with 16.4! 🍾

**Comment by reslear** [comment]
but ok, already work with https://github.com/capawesome-team/capacitor-firebase/tree/main/packages/messaging

**Comment by khromov** [comment]
Is there a way to utilize web push with Capacitor today? It feels suboptimal to require two push implementations (Web Push for PWA and Firebase for Capacitor). Or is it possible to implement the Firebase solution on the web as well so we could have only one solution for all platforms?

**Comment by mleister97** [comment]
Hello! Are there any updates regarding this ticket, or is there a roadmap or timeline available?

**Comment by AvidDabbler** [comment]
Yeah. I would like to know the status of this.

**Comment by marius-se** [comment]
same here!

**Comment by jkscx** [comment]
> Most probably know already, but Web Push Notifications is coming to Safari (both iOS and macOS).
> 
> https://webkit.org/blog/12824/news-from-wwdc-webkit-features-in-safari-16-beta/#web-push-for-macos
> 
> _(I don't know how that would work in a webview app wrapped within a native shell though, which is how most Capacitor apps are packaged on iOS)_

Would be super nice to use WebPush for web/iOS/Androind, but According to an Apple dev currently "[Web Push Notifications will not work in apps with a WKWebView](https://developer.apple.com/forums/thread/760767?login=true)" and I presume that Capacitor is based on WKWebView, so that sadly isn't possible.

Unless the Capacitor team extends the `@capacitor/push-notifications` plugin with WebPush support, we're stuck with the status quo of using WebPush on the web and `@capacitor/push-notifications` on native. The closest to an unified approach to push notifications which I found was the unofficial Capacitor plugin for [Firebase Cloud Messaging](https://firebase.google.com/docs/cloud-messaging), which uses FCM for both web/native, but the setup seems kinda involved.

Would be nice if the Capacitor team could make an abstraction/unification layer on top of WebPush/FCM/APNS, similar to what Expo does with their [Expo Push Service](https://docs.expo.dev/push-notifications/sending-notifications/), but that seems like a big undertaking...
